core/src/main/java/com/linecorp/armeria/client/Http2ResponseDecoder.java

```diff
@@ -96,9 +96,9 @@ private void onWrapperCompleted(HttpResponseWrapper resWrapper, int id, @Nullabl
         resWrapper.onSubscriptionCancelled(cause);
 
         if (cause != null) {
-            // We are not closing the connection but just send a RST_STREAM,
-            // so we have to remove the response manually.
-            removeResponse(id);
```

Comment:
I think we can't only rely on the close events notified by Netty's HTTP2 stream handler.
A request could be failed before writing the first headers if the request fails early.
https://github.com/line/armeria/blob/2920a1ddc2188f150a0d8b77de1d4e2e409997cf/core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java#L169-L172
The early cancelation does not decrease the `unfinishedResponses` and it could be a leak. It rarely happens though.

Reply:
> The early cancelation does not decrease the unfinishedResponses and it could be a leak. It rarely happens though.

Thanks for the detailed review!

I think you're right in that `unfinishedResponses` may not be decremented in the scenario you mentioned.

I'm not sure there will be a leak since the actual `HttpResponseWrapper` isn't created at that point, but I think a connection may not utilize the maximum allowed number of streams. (for instance, although `MAX_CONCURRENT_STREAMS == 5`, the connection may end up using only 3 streams)

I'd like to handle this issue separately in #3920 since this problem slightly diverges from the original issue 🙏

Reply:
Agreed to handle in the f/u PR.